src/coreclr/vm/amd64/Context.asm

```diff
@@ -0,0 +1,106 @@
+        vinsertf128 ymm6, ymm6, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 6 * 16)], 1
+        vinsertf128 ymm7, ymm7, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 7 * 16)], 1
+        vinsertf128 ymm8, ymm8, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 8 * 16)], 1
+        vinsertf128 ymm9, ymm9, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 9 * 16)], 1
+        vinsertf128 ymm10, ymm10, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 10 * 16)], 1
+        vinsertf128 ymm11, ymm11, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 11 * 16)], 1
+        vinsertf128 ymm12, ymm12, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 12 * 16)], 1
+        vinsertf128 ymm13, ymm13, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 13 * 16)], 1
+        vinsertf128 ymm14, ymm14, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 14 * 16)], 1
+        vinsertf128 ymm15, ymm15, xmmword ptr [rcx + (OFFSETOF__CONTEXT__VectorRegister + 15 * 16)], 1
+    Done_Restore_CONTEXT_XSTATE:
+    
+        test    byte ptr [rcx + OFFSETOF__CONTEXT__ContextFlags], CONTEXT_INTEGER
+        je      Done_Restore_CONTEXT_INTEGER_0
+    
+        ; Restore the integer registers except for rax and rcx, which may still be needed and will be restored later
+        mov     rdx, [rcx + OFFSETOF__CONTEXT__Rdx]
+        mov     rbx, [rcx + OFFSETOF__CONTEXT__Rbx]
+        mov     rbp, [rcx + OFFSETOF__CONTEXT__Rbp]
+        mov     rsi, [rcx + OFFSETOF__CONTEXT__Rsi]
+        mov     rdi, [rcx + OFFSETOF__CONTEXT__Rdi]
+        mov     r8, [rcx + OFFSETOF__CONTEXT__R8]
+        mov     r9, [rcx + OFFSETOF__CONTEXT__R9]
+        mov     r10, [rcx + OFFSETOF__CONTEXT__R10]
+        mov     r11, [rcx + OFFSETOF__CONTEXT__R11]
+        mov     r12, [rcx + OFFSETOF__CONTEXT__R12]
+        mov     r13, [rcx + OFFSETOF__CONTEXT__R13]
+        mov     r14, [rcx + OFFSETOF__CONTEXT__R14]
+        mov     r15, [rcx + OFFSETOF__CONTEXT__R15]
+    Done_Restore_CONTEXT_INTEGER_0:
+    
+        test    byte ptr [rcx + OFFSETOF__CONTEXT__ContextFlags], CONTEXT_CONTROL
+        je      Done_Restore_CONTEXT_CONTROL
+    
+        ; When user-mode shadow stacks are enabled, and for example the intent is to continue execution in managed code after
+        ; exception handling, iret and ret can't be used because their shadow stack enforcement would not allow that transition,
+        ; and using them would require writing to the shadow stack, which is not preferable. Instead, iret is partially
+        ; simulated. SS and CS shouldn't be different from their current values in cases where this function is used.
+        mov     rsp, [rcx + OFFSETOF__CONTEXT__Rsp]     ; restore rsp
+        mov     eax, [rcx + OFFSETOF__CONTEXT__EFlags]  ; push RFlags onto the stack
+        push    rax
+        mov     rax, [rcx + OFFSETOF__CONTEXT__Rip]     ; store the target IP on the stack below RFlags
+        mov     [rsp - 8h], rax
```

Comment:
I don't think we can write below RSP on Windows. I believe Windows doesn't support red zone and if interrupt occured or APC got injected, the memory could get corrupted.

Reply:
I see, good point. In that case I don't see a way to do this without writing to the shadow stack, any ideas?